node/service/src/lazy_loading/mod.rs

```diff
@@ -665,17 +672,39 @@ where
 							));
 						}
 
+						let current_para_head = client_for_xcm
+							.header(block)
+							.expect("Header lookup should succeed")
+							.expect("Header passed in as parent should be present in backend.");
+
+						let should_send_go_ahead = match client_for_xcm
+							.runtime_api()
+							.collect_collation_info(block, &current_para_head)
+						{
+							Ok(info) => info.new_validation_code.is_some(),
+							Err(e) => {
+								log::error!("Failed to collect collation info: {:?}", e);
+								false
+							}
+						};
+
 						let mocked_parachain = MockValidationDataInherentDataProvider {
 							current_para_block,
 							para_id: ParaId::new(parachain_id),
+							upgrade_go_ahead: should_send_go_ahead.then(|| {
```

Comment:
This means that we can remove the `has_pending_upgrade` and `upgrade_go_ahead_signal` logic above, this `upgrade_go_ahead` will do it internally.

Reply:
Addressed in https://github.com/moonbeam-foundation/moonbeam/pull/3224/commits/43d380c0f676fa86d5e6eddf7564f3900d67d21f Please, double check.